### DISCUSSION TITLE
Changing device config in runtime

### DISCUSSION THREAD

**Discussion by robertkowalewski** [zephyrproject-rtos/zephyr#61406] [Q&A] [answered]
Hi,
I have a case where I need to update **cs.delay** (from SPI driver) value for wakeup frame sent to SPI slave device (for other messages **cs.delay** will have its default value which is 0). As _device config_ is read-only structure, I cannot update that property in runtime. So I came up with an idea to allocate a "new" device structure and copy the origin device structure into it, and then changed a value of **cs.delay**. Does it have a chance to work? Is there other way to achieve the same goal?

**Example:**
```
struct device *tmp_dev = k_malloc(sizeof(struct device));

tmp_dev = k_malloc(sizeof(struct device));
memcpy(tmp_dev, dev, sizeof(struct device));

tmp_config = dev->config;
tmp_config->bus.config.cs.delay = 400;

spi_write_dt(&tmp_config->bus, &tx);
```

**Comment by yashi** [comment, marked as answer]
If you are not using the values from devicetree, you can just use plain i2c functions (without `_dt` suffix).
For `spi_write_dt()`, there is [`spi_write()`](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/spi.h#L830-L835), which takes `dev`, `config`, and `tx_bufs`.  `struct spi_config`, or `spi_cs_control` within it, is what you need to check.

https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/spi.h#L284-L315

You can pass your spi_config to your spi device how to operate.

**Reply by robertkowalewski** [reply]
Thanks for your answer, it was really helpful!
